enhancements/local-storage/automatic-detection.md

```diff
@@ -0,0 +1,161 @@
+- Should respond to the attach/detach events of the disks/devices.
+- Should have options for filtering particular kind of disks based on properties such as name, size, manufacturer, etc.
+
+### Non-Goals
+
+## Proposal
+
+`local-storage-operator`is already capable of consuming local disks from the nodes and provisioning PVs out of it, but the disk/device paths needs to explicitly specified in the `LocalVolume` CR. The idea is to bring in one more layer of CR above it called `AutoDetectVolumes`.
+The controller for this CR will be responsible for automatically detecting disks/devices from the available nodes and creating/managing the life cycle of `LocalVolumes` which are governed by the `AutoDetectVolumes`.s
+
+### Risks and Mitigations
+
+- Detecting disks that contain data and/or are in-use.
+- Ensuring disks aren't re-detected as new or otherwise destroyed if their device path changes.
+
+## Design Details
+
+API scheme for `AutoDetectVolumes`:
+
+```
+type DeviceDiscoveryPolicyType string
+
+const (
+	raid DeviceDiscoveryPolicyType = "raid"
+	disk DeviceDiscoveryPolicyType = "disk"
+)
+
+type AutoDetectVolumeList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+	Items           []AutoDetectVolume `json:"items"`
+}
+
+type AutoDetectVolume struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata"`
+	Spec              AutoDetectVolumeSpec   `json:"spec"`
+	Status            AutoDetectVolumeStatus `json:"status,omitempty"`
+}
+
+type AutoDetectVolumeSpec struct {
+	// StorageClass name to use for set of matched devices
+	StorageClassName string `json:"storageClassName"`
+	// Volume mode. Raw or with file system
+	// + optional
+	VolumeMode PersistentVolumeMode `json:"volumeMode,omitempty"`
+	// File system type
+	// +optional
+	FSType string `json:"fsType,omitempty"`
+	// Nodes on which the autoDetection must run
+	// +optional
+	NodeSelector *corev1.NodeSelector `json:"nodeSelector,omitempty"`
+	// DeviceDiscoverPolicies are the list of policies that AutoDetectVolume's controller
+	// will be looking for during detection
+	DeviceDiscoveryPolicies []DeviceDiscoveryPolicy `json:"deviceDiscoveryPolicies,omitempty"`
+}
+
+type AutoDetectVolumeStatus struct {
+	Status string `json:"status"`
+}
+
+type DeviceDiscoveryPolicy struct {
+	// Device type that should be used for auto detection. This would be one of the types supported
+	// by the local-storage operator. Initial possible types can be - crypt, raid1, raid4, raid5,
+	// raid10, multipath, disk, tape, printer, processor, worm, rom, scanner, mo-disk, changer,
+	// comm, raid, enclosure, rbc, osd, and no-lun.
+	DeviceType DeviceDiscoveryPolicyType `json:"deviceType"`
+	// A list of regular expressions that will be used to exclude certain devices
+	// For example - ["^rbd[0-9]+p?[0-9]{0,}$"]
+	// +optional
+	DeviceInclusionFilter []string `json:"deviceInclusionFilter"`
```

Comment:
Why is this an array? `DeviceDiscoveryPolicies` is already an array and `DeviceType` is a single value but `DeviceInclusionFilter` is an array. This does not make much sense.

Reply:
Changed it to string.